librz/util/float/float.c

```diff
@@ -0,0 +1,1288 @@
+	rz_bv_free(sig);
+	free(exp_s);
+	free(m_s);
+}
+
+/*
+ * Common NaN and Inf detection
+ * */
+#define PROC_SPECIAL_FLOAT_START(left, right) \
+	{ \
+		RzFloatSpec l_type, r_type; \
+		l_type = rz_float_detect_spec((left)); \
+		r_type = rz_float_detect_spec((right)); \
+		bool l_is_inf = (l_type == RZ_FLOAT_SPEC_PINF || l_type == RZ_FLOAT_SPEC_NINF); \
+		bool r_is_inf = (r_type == RZ_FLOAT_SPEC_PINF || r_type == RZ_FLOAT_SPEC_NINF); \
+		bool l_is_nan = (l_type == RZ_FLOAT_SPEC_SNAN || l_type == RZ_FLOAT_SPEC_QNAN); \
+		bool r_is_nan = (r_type == RZ_FLOAT_SPEC_SNAN || r_type == RZ_FLOAT_SPEC_QNAN); \
+		bool l_is_zero = l_type == RZ_FLOAT_SPEC_ZERO; \
+		bool r_is_zero = r_type == RZ_FLOAT_SPEC_ZERO;
+
+#define PROC_SPECIAL_FLOAT_END }
+
+/**
+ * Get const attribute from float
+ * @param format RzFloatFormat, format of a float
+ * @param which_info Specify an attribute
+ * @return ut32 const value bind with `which_info`
+ */
+RZ_API ut32 rz_float_get_format_info(RzFloatFormat format, RzFloatInfo which_info) {
+	switch (format) {
+	case RZ_FLOAT_IEEE754_BIN_32:
+		return rz_float_info_bin32(which_info);
+	case RZ_FLOAT_IEEE754_BIN_64:
+		return rz_float_info_bin64(which_info);
+	case RZ_FLOAT_IEEE754_BIN_128:
+		return rz_float_info_bin128(which_info);
+	case RZ_FLOAT_IEEE754_DEC_64:
+	case RZ_FLOAT_IEEE754_DEC_128:
+	default:
+		RZ_LOG_ERROR("TODO");
+		return 0;
+	}
+}
+
+/**
+ * Finish the bv inside the float, and set all to NULL
+ * @param f float
+ */
+RZ_API void rz_float_fini(RZ_NONNULL RzFloat *f) {
+	rz_return_if_fail(f);
+	rz_bv_free(f->s);
+	memset(f, 0, sizeof(RzFloat));
+}
+
+/**
+ * Destroy the float structure
+ * @param f float
+ */
+RZ_API void rz_float_free(RZ_NULLABLE RzFloat *f) {
+	if (!f) {
+		return;
+	}
+	rz_float_fini(f);
+	free(f);
+}
+
+/**
+ * Init the bv inside float
+ * @param f float
+ * @return return true if init success else return false
+ */
+RZ_API bool rz_float_init(RZ_NONNULL RzFloat *f) {
+	rz_return_val_if_fail(f, false);
+	rz_float_fini(f);
+
+	RzFloatFormat format = f->r;
+	ut32 exp_len = rz_float_get_format_info(format, RZ_FLOAT_INFO_EXP_LEN);
+	ut32 man_len = rz_float_get_format_info(format, RZ_FLOAT_INFO_MAN_LEN);
+
+	f->s = rz_bv_new(1 + exp_len + man_len);
+	if (!f->s) {
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * Create float and init it
+ * @param format float format
+ * @return return an RzFloat instance with zero value
+ */
+RZ_API RZ_OWN RzFloat *rz_float_new(RzFloatFormat format) {
+	RzFloat *f = RZ_NEW(RzFloat);
+	if (!f) {
+		return NULL;
+	}
+	f->r = format;
+	f->s = NULL;
+
+	if (!rz_float_init(f)) {
+		free(f);
+		return NULL;
+	}
+
+	return f;
+}
+
+/**
+ * Duplicate a float
+ * @param f float
+ * @return a copy of float
+ */
+RZ_API RZ_OWN RzFloat *rz_float_dup(RZ_NONNULL RzFloat *f) {
+	rz_return_val_if_fail(f, NULL);
+	RzFloat *cp = RZ_NEW(RzFloat);
+	if (!cp) {
+		RZ_LOG_ERROR("Dup float failed")
+		return NULL;
+	}
+
+	cp->r = f->r;
+	cp->s = rz_bv_dup(f->s);
+
```

Comment:
Not quite sure about this: But I would also make a copy of `f->exception` although it might be very unlikely that `rz_loat_dup` is called after an exception was set.